File: P1_2DMatrix.cpp
//Time Complexity: O(log(n)) + O(log(m)), binary search (rows) and binary search (columns)
//Space Complexity: O(1), no extra space used
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach


class Solution {
public:
    int findRow(vector<vector<int>>& matrix, int low, int high, int tar){
        int mid = low;
        while(low<high) {
            mid = low + (high-low)/2;
            if(tar == matrix[mid][0]){
                return mid;
            } else if(tar < matrix[mid][0]){
                high = mid-1;
            } else { //tar > matrix[mid][0]
                if(tar < matrix[mid+1][0]){
                    return mid;
                }
                low = mid+1;
            }
        }
        return low;
    }
    
    bool findColumn(vector<vector<int>>& matrix, int row, int low, int high, int tar){
        while(low<=high){
            int mid = low + (high-low)/2;
            if(matrix[row][mid] == tar){
                return true;
            } else if(matrix[row][mid] < tar){
                low = mid+1;
            } else {
                high = mid - 1;
            }
        }
        return false;
    }
    
    
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rcount = matrix.size();
        int ccount = matrix[0].size();
        //cout << rcount << ", col:" << ccount << endl;
        int row = findRow(matrix, 0, rcount-1, target);
        bool result = findColumn(matrix, row, 0, ccount-1, target);
        return result;
    }
};

File: p1_2DMatrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        # check if the matrix exist
        if not matrix or not matrix[0]:
            return False

        #setting rows, column and left,right pointers
        m, n = len(matrix), len(matrix[0])
        left, right = 0, m * n - 1

        while left <= right:
            # finding middle index and pull the value
            mid = left + (right - left) // 2
            mid_element = matrix[mid // n][mid % n]

            if mid_element == target:
                return True
            
            #compare to update pointer for next search
            elif mid_element < target:
                left = mid + 1
            else:
                right = mid - 1
        
        # return false if search fails
        return False

# ------------------------------------------------------------------------------------------------------------------------
# Time complexity: O(logn)